File: android/src/main/java/me/carda/awesome_notifications/background/DartBackgroundService.java
package me.carda.awesome_notifications.background;

import android.content.Context;
import android.content.Intent;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.core.app.JobIntentService;
import me.carda.awesome_notifications.notifications.managers.DefaultsManager;

public class DartBackgroundService extends JobIntentService {
    private static final String TAG = "DartBackgroundService";

    @Override
    protected void onHandleWork(@NonNull final Intent intent) {

        long dartCallbackHandle = getDartCallbackDispatcher(this);
        if (dartCallbackHandle == 0L) {
            Log.w(TAG, "A background message could not be handled in Dart" +
                            " because there is no onSilentData handler registered.");
            return;
        }

        long silentCallbackHandle = getSilentCallbackDispatcher(this);
        if (silentCallbackHandle == 0L) {
            Log.w(TAG,"A background message could not be handled in Dart" +
                            " because there is no dart background handler registered.");
            return;
        }

        DartBackgroundExecutor.runBackgroundExecutor(
                this,
                intent,
                dartCallbackHandle,
                silentCallbackHandle);
    }

    public static long getDartCallbackDispatcher(Context context){
        return DefaultsManager.getDartCallbackDispatcher(context);
    }

    public static long getSilentCallbackDispatcher(Context context){
        return DefaultsManager.getSilentCallbackDispatcher(context);
    }
}

File: ios/Classes/lib/notifications/background/DartBackgroundService.swift
//
//  DartBackgroundServices.swift
//  awesome_notifications
//
//  Created by Rafael Setragni on 07/06/21.
//

import Foundation

class DartBackgroundService {
    private static let TAG:String = "DartBackgroundService"
    
    private static var instance:DartBackgroundService?
    static var shared: DartBackgroundService = {
        instance = instance ?? DartBackgroundService()
        return instance!
    }()

    private init() {}
    
    public static func executeAction(
        actionReceived:ActionReceived,
        handler: @escaping () -> ()
    ){
        Log.d(TAG, "A new Dart background service has started")
        
        let silentActionRequest:SilentActionRequest = SilentActionRequest(
            actionReceived: actionReceived,
            handler: handler
        )
        
        let dartCallbackHandle: Int64 = DefaultManager.getDartBgCallback()
        let silentCallbackHandle: Int64 = DefaultManager.getActionCallback()
        
        if dartCallbackHandle == 0 {
            Log.d(DartBackgroundService.TAG,
                  "A background message could not be handled in Dart because there is no valid onActionReceivedMethod handler register")
            return
        }
        
        if silentCallbackHandle == 0 {
            Log.d(DartBackgroundService.TAG,
                  "A background message could not be handled in Dart because there is no valid dart background handler register"
            )
            return
        }
        
        DartBackgroundExecutor.shared.runBackgroundExecutor(
            silentActionRequest: silentActionRequest,
            dartCallbackHandle: dartCallbackHandle,
            silentCallbackHandle: silentCallbackHandle
        )
    }
}
